exchanges/gateio/gateio.go

```diff
@@ -420,170 +1025,2884 @@ func (g *Gateio) SendAuthenticatedHTTPRequest(ctx context.Context, ep exchange.U
+}
+
+// GetCrossMarginBorrowHistory retrieves cross margin borrow history sorted by creation time in descending order by default.
+// Set reverse=false to return ascending results.
+func (g *Gateio) GetCrossMarginBorrowHistory(ctx context.Context, status uint64, ccy currency.Code, limit, offset uint64, reverse bool) ([]CrossMarginLoanResponse, error) {
+	if status < 1 || status > 3 {
+		return nil, fmt.Errorf("%s %v, only allowed status values are 1:failed, 2:borrowed, and 3:repayment", g.Name, errInvalidOrderStatus)
+	}
+	params := url.Values{}
+	params.Set("status", strconv.FormatUint(status, 10))
+	if !ccy.IsEmpty() {
+		params.Set("currency", ccy.String())
+	}
+	if limit > 0 {
+		params.Set("limit", strconv.FormatUint(limit, 10))
+	}
+	if offset > 0 {
+		params.Set("offset", strconv.FormatUint(offset, 10))
+	}
+	if reverse {
+		params.Set("reverse", strconv.FormatBool(reverse))
+	}
+	var response []CrossMarginLoanResponse
+	return response, g.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, spotPrivateEPL, http.MethodGet, crossMarginLoans, params, nil, &response)
+}
+
+// GetSingleBorrowLoanDetail retrieve single borrow loan detail
+func (g *Gateio) GetSingleBorrowLoanDetail(ctx context.Context, loanID string) (*CrossMarginLoanResponse, error) {
+	if loanID == "" {
+		return nil, errInvalidLoanID
+	}
+	var response *CrossMarginLoanResponse
+	return response, g.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, spotPrivateEPL, http.MethodGet, crossMarginLoans+"/"+loanID, nil, nil, &response)
+}
+
+// *********************************Futures***************************************
+
+// GetAllFutureContracts  retrieves list all futures contracts
+func (g *Gateio) GetAllFutureContracts(ctx context.Context, settle string) ([]FuturesContract, error) {
+	settle = strings.ToLower(settle)
+	if settle != settleBTC && settle != settleUSD && settle != settleUSDT {
+		return nil, errInvalidSettleCurrency
+	}
+	var contracts []FuturesContract
+	return contracts, g.SendHTTPRequest(ctx, exchange.RestSpot, perpetualSwapDefaultEPL, "futures/"+settle+"/contracts", &contracts)
+}
+
+// GetSingleContract returns a single contract info for the specified settle and Currency Pair (contract << in this case)
+func (g *Gateio) GetSingleContract(ctx context.Context, settle, contract string) (*FuturesContract, error) {
+	if contract == "" {
+		return nil, currency.ErrCurrencyPairEmpty
+	}
+	settle = strings.ToLower(settle)
+	if settle != settleBTC && settle != settleUSD && settle != settleUSDT {
+		return nil, errInvalidSettleCurrency
+	}
+	var futureContract *FuturesContract
+	return futureContract, g.SendHTTPRequest(ctx, exchange.RestSpot, perpetualSwapDefaultEPL, fmt.Sprintf(futuresSettleContracts+"/%s", settle, contract), &futureContract)
+}
+
+// GetFuturesOrderbook retrieves futures order book data
+func (g *Gateio) GetFuturesOrderbook(ctx context.Context, settle, contract, interval string, limit uint64, withOrderbookID bool) (*Orderbook, error) {
```

Comment:
After running GateIO, we get the futures currency `ETH_USD`, however, when calling this function it returns `400 raw response: {"label":"CONTRACT_NOT_FOUND"}`
I think there's an issue with either this implementation, or you're putting the wrong currencies in the wrong asset types